Refactor missused ViewModels, support automated UI tests and screenshots (EXPOSUREAPP-2946)

d4rken opened the pull request:
## Why is this PR so huge? :fearful:
Ripping the :adhesive_bandage:  off in one swoop saves us weeks of work.

The core idea is that we want to instantiate screens on their own to automate UI tests and take automated screenshots. This was impossible because there were too many dependencies that would have to be satisfied for each screen, sepperation was not possible because each screen accessed various other components via static access.

We had a lot of :ramen: code based on (`TracingViewModel`, `SubmissionViewModel` and `SettingsViewModel`) that was very interwoven in the apps structure. This PR removes the :ramen: and creates a one-way information flow. By that I mean that information flows from internal components (i.e. Bluetooth state) to external components (Fragments showing something) without extra roundtrips (i.e. we don't want individual fragments calling `refreshBluetooth`).

## What changed
* Reactive information flow for many elements, i.e. UI subscribes data, data changes, UI updates automatically, the UI no longer manually calls refresh in most cases.
* Fewer singleton access to allow easier testing.
* Most `***FormaterHelper` calls have been removed from the layout files, we had too much logic happening in XML files. Too difficult to add new features, too easy to break existing logic.
* We now have UI "state" objects. The viewmodels have a `Flow<State>` that can be observed, and if it changes the Fragment will update the layout via databinding. The remaining databinding code just sets the state variables to the relevant views. So now we have the best of both worlds. Not too complex logic in XML code. No extra boilerplate code in Fragments.
    * The states can easily be unit-tested on their own.
    * When writing instrumentation tests or doing  automated screenshots, we only need to create mock state objects to make the UI look a specific way.

The above patterns are something we should use throughout the app. This PR already refactors a lot but does not refactor everything. Only the places that were connected to the classes that needed to be refactored in relation to the `HomeFragment`... which is a lot :sweat_smile: 

## Testing
* Check the notification settings behavior in different states
* Check the tracing settings behavior in different states
* Check the behavior of the tracing status headline on the home screen in different states
* Check the risk card behavior in different states
* Check the submission card behavior in different states
* Check the risk details screen in different states
* Check submission flow

d4rken committed:
Introduced `TracingCardState` and `TracingDetailsState`.
Removed all direct viewmodel/livedata access from all nested layouts for `TracingViewModel`.
Removed `FormatterRiskHelper`.
Introduced logic to reactively update the UI state if bluetooth, location or tracing state changes.

TODO:
* Finish tests in TracingCardStateTest.kt and TracingDetailsStateTest.kt
* Finish TracingDetailsViewModel.kt and TracingCardViewModel.kt
* Remove obsolete dependencies on TracingViewModel.kt and singletons RiskLevelRepository.kt, TracingRepository.kt

d4rken committed:
Breakthrough, cards update again.

d4rken committed:
Further cleanup of TracingViewModel.kt, TracingRepository.kt
Still not quite there yet, too many calls going into every direction, back and forth, instead of just top->down, but we are getting there.
We want data flow from business logic code to the UI, keep the UI KISS.

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/TracingRepository.kt:
```diff
@@ -23,21 +24,26 @@ object TracingRepository {
+            LocalData.lastTimeDiagnosisKeysFromServerFetch()
     }
 
+    // TODO shouldn't access this directly
```
Should go to top of class

ralfgehrer committed:
Update TracingCardState tests to account for changes

d4rken committed:
More refactoring, TracingViewModel.kt is now finally gone.

d4rken committed:
Merge remote-tracking branch 'origin/fix/2946-refactor-viewmodels' into fix/2946-refactor-viewmodels

# Conflicts:
#	Corona-Warn-App/src/test/java/de/rki/coronawarnapp/ui/tracing/card/TracingCardStateTest.kt

d4rken committed:
Merge branch 'release/1.6.x' into fix/2946-refactor-viewmodels

d4rken committed:
Fix risk details card, we reuse the cardviewmodel but need to overwrite the `showDetails` flag depending from where it's used.

d4rken committed:
Refactored several properties (notification related) out of SettingsViewModel/Repository

d4rken committed:
Fix injection

d4rken committed:
Refactoring submission related databinding code on the home fragment

d4rken committed:
Make TracingRepository.kt injectable and change viewmodel scoping for refresh calls to app scope.

d4rken committed:
Remove additional formatter calls from submission cards on home screen.

d4rken committed:
Refactored additional SubmissionViewModel logic into SubmissionRepository on the Homescreen.
First UI Test is now passing.

Simplified ViewModel instantiation related code.
Prepared HomeFragment for smoother UI state changes, less reloading.

d4rken committed:
Improved home,tracing-details and tracing-settings loading behavior, less flickering.
Fixed LiveData value assignment on background-thread.

d4rken committed:
New reactive provider for Bluetooth and Location status.
Cleaned up ConnectivityHelper.kt partially.

d4rken committed:
Merge branch 'release/1.6.x' into fix/2946-refactor-viewmodels

# Conflicts:
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/MainActivityModule.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/fragment/SubmissionResultPositiveOtherWarningFragment.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/fragment/SubmissionTestResultFragment.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt

d4rken committed:
Implement remaining tests, batch 1.

chris-cwa reviewed (dismissed):
This code is so clean you could perform open heart surgery on it!

d4rken committed:
Implement remaining tests, batch 2.

d4rken committed:
Fix lint issues.

d4rken committed:
Removed a few codesmells and code duplication.
Introduced Flow.shareLatest(...)

d4rken committed:
Removed unused parameters.

d4rken committed:
LINT fixes.

ralfgehrer commented on Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/api/ui/TestForAPIFragment.kt:
```diff
@@ -258,34 +250,40 @@ class TestForAPIFragment : Fragment(R.layout.fragment_test_for_a_p_i),
+                vm.launch {
+                    val daysInRetention =
+                        TimeVariables.getActiveTracingDaysInRetentionPeriod().toString()
+                    withContext(Dispatchers.Main) {
```
Just a test fragment but maybe you want to inject the dispatcher provider anyway

d4rken replied:
The test fragment is not being tested currently, when we move this logic to a viewmodel, then we should use the injected dispatcher. While it's better, doesn't offer us much to do it now until it's under test.

ralfgehrer commented on Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/api/ui/TestForAPIFragment.kt:
```diff
@@ -258,34 +250,40 @@ class TestForAPIFragment : Fragment(R.layout.fragment_test_for_a_p_i),
+                    val intervals = TracingIntervalRepository.getDateRepository(requireContext())
+                        .getIntervals()
+                        .toString()
+                    withContext(Dispatchers.Main) {
```
Just a test fragment but maybe you want to inject the dispatcher provider anyway

ralfgehrer commented on Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/api/ui/TestForAPIFragment.kt:
```diff
@@ -258,34 +250,40 @@ class TestForAPIFragment : Fragment(R.layout.fragment_test_for_a_p_i),
+                    val summary = ExposureSummaryRepository.getExposureSummaryRepository()
+                        .getExposureSummaryEntities().toString()
+
+                    withContext(Dispatchers.Main) {
```
Just a test fragment but maybe you want to inject the dispatcher provider anyway

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SettingsRepository.kt:
```diff
@@ -23,65 +25,12 @@ class SettingsRepository @Inject constructor(
-    }
+    private val internalIsBackgroundPriorityEnabled = MutableStateFlow(false)
+    val isBackgroundPriorityEnabledFlow: Flow<Boolean> = internalIsBackgroundPriorityEnabled
+    val isBackgroundPriorityEnabled = internalIsBackgroundPriorityEnabled.asLiveData()
```
I suppose the live data one will be deprecated eventually and you don't rename it to not change a million other occurrences? I'd propose to mark it as deprecated at least with a comment then.

d4rken replied:
Ohh I like the idea, adding  `@Deprecated` :+1:

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SettingsRepository.kt:
```diff
@@ -119,24 +50,33 @@ class SettingsRepository @Inject constructor(
 
+    private val internalIsManualKeyRetrievalEnabled = MutableStateFlow(true)
+    val isManualKeyRetrievalEnabledFlow: Flow<Boolean> = internalIsManualKeyRetrievalEnabled
+    val isManualKeyRetrievalEnabled = isManualKeyRetrievalEnabledFlow.asLiveData()
```
same as above and for others in this repo

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/tracing/settings/SettingsTracingFragment.kt:
```diff
@@ -141,7 +137,7 @@ class SettingsTracingFragment : Fragment(R.layout.fragment_settings_tracing),
                 if (InternalExposureNotificationClient.asyncIsEnabled()) {
                     InternalExposureNotificationClient.asyncStop()
-                    tracingViewModel.refreshIsTracingEnabled()
+//                    tracingViewModel.refreshIsTracingEnabled()
```
If intended, why not delete it as with others.

d4rken replied:
Was a reminder while refactoring, removed :+1:

ralfgehrer reviewed (dismissed):
works well on my device (Pixel 1, Android 10)

BMItter commented:
Tested on Pixel 2 (A11) & Nexus 5x (A8.1):
risk card & states, low risk tested. **i.o,** unknown, **i.o**
tracing state behavior **i.o.**, UI updates accordingly, **i.o**
tracing status headline on home screen, **i.o**
submission of test and UI check, states: pending **i.o**, negativ **i.o**, positiv **i.o.** , no test **i.o**
tested DefaultScanningSupport (new singleton), **i.o.**
Still doing some checks.

One short note: if risk detection is stopped by switch, after app restart initial card is showing "unknown risk", jumping to "riskdetection stopped" card. It might be the default behaviour in lack of peristancy, but I want to mention this.

d4rken committed:
Fix unit test failing due to missing mocks.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt:
```diff
@@ -58,27 +60,75 @@ object SubmissionRepository {
         LocalData.teletan(teletan)
     }
+
+    private val _uiStateError = MutableLiveData<Event<CwaWebException>>(null)
```
Was copy pasted from previous code, but can fix.

d4rken committed:
Address PR comments.

d4rken committed:
Improve UI loading states, remove flickering.

d4rken committed:
Remove log spam due to polling values, add extra abort condition that checks for cancellation.
